ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,501 @@
+      // notify account changes after successfully update.
+      if (notifier.publish(ACCOUNT_METADATA_CHANGE_TOPIC, FULL_ACCOUNT_METADATA_CHANGE_MESSAGE)) {
+        logger.trace("Successfully published message for account metadata change");
+      } else {
+        logger.error("Failed to send notification for account metadata change");
+        accountServiceMetrics.notifyAccountDataChangeErrorCount.inc();
+      }
+    } else {
+      logger.error("Failed updating accounts={}", accounts);
+      accountServiceMetrics.updateAccountErrorCount.inc();
+    }
+    return hasSucceeded;
+  }
+
+  @Override
+  public void close() {
+    if (isOpen) {
+      try {
+        isOpen = false;
+        helixStore.stop();
+      } catch (Exception e) {
+        logger.error("Exception occurred when closing HelixAccountService.", e);
+      }
+    }
+  }
+
+  /**
+   * Reads the full set of {@link Account} metadata from {@link HelixPropertyStore}, and update the local cache.
+   *
+   * @param pathToFullAccountMetadata The path to read the full set of {@link Account} metadata.
+   */
+  private void readFullAccountAndUpdateCache(String pathToFullAccountMetadata) throws JSONException {
+    lock.lock();
+    try {
+      long startTimeMs = System.currentTimeMillis();
+      logger.trace("Start reading full account metadata set from path={}", pathToFullAccountMetadata);
+      ZNRecord zNRecord = helixStore.get(pathToFullAccountMetadata, null, AccessOption.PERSISTENT);
+      logger.trace("Fetched ZNRecord from path={}, took time={} ms", pathToFullAccountMetadata,
+          System.currentTimeMillis() - startTimeMs);
+      if (zNRecord == null) {
+        logger.debug("The ZNRecord to read does not exist on path={}", pathToFullAccountMetadata);
+      } else {
+        Map<String, String> remoteAccountMap = zNRecord.getMapField(ACCOUNT_METADATA_MAP_KEY);
+        if (remoteAccountMap == null) {
+          logger.debug("ZNRecord={} to read on path={} does not have a simple map with key={}", zNRecord,
+              pathToFullAccountMetadata, ACCOUNT_METADATA_MAP_KEY);
+        } else {
+          logger.trace("Start parsing remote account data.");
+          AccountInfoMap newAccountInfoMap = new AccountInfoMap(remoteAccountMap);
+          accountInfoMapRef.set(newAccountInfoMap);
+          logger.trace("Completed updating local cache with number={} accounts",
+              newAccountInfoMap.getAccounts().size());
+        }
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Checks if there are duplicate accountId or accountName in the given collection of {@link Account}s.
+   *
+   * @param accounts A collection of {@link Account}s to check duplication.
+   * @return {@code true} if there are duplicated accounts in id or name in the given collection of accounts,
+   *                      {@code false} otherwise.
+   */
+  private boolean hasDuplicateAccountIdOrName(Collection<Account> accounts) {
+    if (accounts == null) {
```

Comment:
I am a bit inclined to keep this, because this is not really a null check, but part of the method's logic. It basically means a `null` collection of `Account`s do not have conflicting `Account`.